### PR TITLE
jsre: support for setTimeout, sendAsync 

### PR DESCRIPTION
**Pull request by zsfelfoldi**
using a modified version of https://github.com/robertkrimen/natto/blob/master/natto.go
also removed provider polling from Ethereum_JS
added natto test to jsre_test

### CODE REVIEW

**Comment by zsfelfoldi** [comment]
https://github.com/ethereum/go-ethereum/issues/772

File: jsre/natto/natto.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+// Run will execute the given JavaScript, continuing to run until all timers have finished executing (if any).
+// The VM has the following functions available:
+//
+//      <timer> = setTimeout(<function>, <delay>, [<arguments...>])
+//      <timer> = setInterval(<function>, <delay>, [<arguments...>])
+//      clearTimeout(<timer>)
+//      clearInterval(<timer>)
+//
+func Run(vm *otto.Otto, src interface{}) (value otto.Value, err error) {
```
Comment (fjl) [inline jsre/natto/natto.go]: I don't see a reason why this should be in a separate package. Please move this to package `jsre`.

Comment (zsfelfoldi) [inline jsre/natto/natto.go]: done

File: jsre/natto.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+This file is a modified version of https://github.com/robertkrimen/natto/blob/master/natto.go
+
+Package natto is an example/offshoot of otto that implements an event loop (supporting setTimeout/setInterval).
+
+http://godoc.org/github.com/robertkrimen/natto
+
+otto: http://github.com/robertkrimen/otto (http://godoc.org/github.com/robertkrimen/otto)
+
+*/
```
Comment (fjl) [inline jsre/natto.go]: Please remove this comment.

File: jsre/natto.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+				}
+			}
+			if timer.interval {
+				timer.timer.Reset(timer.duration)
+			} else {
+				delete(registry, timer)
+			}
+		default:
+			// Escape valve!
+			// If this isn't here, we deadlock...
```
Comment (fjl) [inline jsre/natto.go]: Can you explain why the default case is necessary? I don't see it.

File: jsre/natto.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+		}
+		if len(registry) == 0 {
```
Comment (fjl) [inline jsre/natto.go]: This condition should be on the `for` statement instead.

File: jsre/natto.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+*/
+package jsre
+
+import (
+	"github.com/robertkrimen/otto"
+	"time"
+)
+
+type _timer struct {
```
Comment (fjl) [inline jsre/natto.go]: Please rename this to `timer`.

File: jsre/natto.go (hunk 2 above)
Comment (fjl) [inline jsre/natto.go]: Please remove this comment.

File: jsre/natto.go (hunk 3 above)
Comment (zsfelfoldi) [inline jsre/natto.go]: That was part of the original https://github.com/robertkrimen/natto/blob/master/natto.go
I only did minimal necessary modifications.

**Comment by fjl** [comment]
I think this doesn't quite do what we want in the console. The console works like this:

```go
    code := readLine()
    result := jsre.Run(code)
    print(result)
```

With the 'natto' approach so far, the second step would simply hang if the JS code installed any timers.
In the console, we need an approach where code can be injected into the event loop and where the event loop can be aborted at any time. This can be done by structuring the loop as follows:

```go
type evalReq struct {
	code string
	result *otto.Value
	err  error
	done chan struct{}
}

loop:
for {
    select {
	case timer := <-timerReady:
		// execute callback, remove/reschedule the timer
	case evalReq := <-eval:
		// run the code, send the result back
        evalReq.result, evalReq.err = vm.Run(evalReq.code)
		close(evalReq.done)
	case <-stop:
		break loop
    }
}
```

**Comment by fjl** [comment]
I think the proper fix for ethereum.js would be to ensure that it removes all of its timers and intervals when all filters are deregistered. This would provide a way for the application to exit cleanly. It also needs to do that for node.js.

File: jsre/natto.go [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+// The VM has the following functions available:
+//
+//      <timer> = setTimeout(<function>, <delay>, [<arguments...>])
+//      <timer> = setInterval(<function>, <delay>, [<arguments...>])
+//      clearTimeout(<timer>)
+//      clearInterval(<timer>)
+//
+
+*/
```
Comment (fjl) [inline jsre/natto.go]: When writing "please remove this comment", I meant "please remove the package comment".

File: jsre/natto.go [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+type natto struct {
+	vm      *otto.Otto
+	eval    chan *evalReq
+	stop    chan bool
+	running bool
+}
+
+func newNatto(vm *otto.Otto) (self *natto) {
```
Comment (fjl) [inline jsre/natto.go]: It is kind-of pointless to have this as a separate type. Please move the logic into `JSRE`.

File: jsre/natto.go [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+				timer.timer.Reset(timer.duration)
+			} else {
+				delete(registry, timer)
+			}
+		case evalReq := <-self.eval:
+			// run the code, send the result back
+			self.vm.Set("setTimeout", setTimeout)
+			self.vm.Set("setInterval", setInterval)
+			self.vm.Set("clearTimeout", clearTimeout)
+			self.vm.Set("clearInterval", clearTimeout)
```
Comment (fjl) [inline jsre/natto.go]: When you move the logic into `JSRE`, you can register these during initialization.

File: jsre/natto.go [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+			close(evalReq.done)
+		case <-self.stop:
+			break loop
+		}
+	}
+}
+
+func (self *natto) run(src interface{}) (value otto.Value, err error) {
+
+	if self.running {
```
Comment (fjl) [inline jsre/natto.go]: What does this check do?

File: jsre/natto.go [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+		done: done,
+	}
+	self.eval <- req
+	<-done
+	self.running = false
+	return req.result, req.err
+}
+
+func (self *natto) stopEventLoop() {
+	close(self.stop)
```
Comment (fjl) [inline jsre/natto.go]: This is missing shutdown synchronization (e.g. with `sync.WaitGroup`). You don't know whether the loop has actually shut down in this case.

**Comment by fjl** [comment]
This is still missing the 'extended' stop logic for file execution.

Here is an example you can try both in the console and when executing from a file:

```js
function callback(n) {
    console.log("callback", n)
    if (n > 0) {
        setTimeout(callback, 1, n - 1);
    }
}

setTimeout(callback, 1, 10);
```

When you run this in `geth console`, it should be possible to exit the console immediately by typing `exit<Return>` or `^D`. The console should not wait for the callbacks.

When running with `geth js test.js` (assuming you put the above code into `test.js`), the program should exit after 10 seconds.

**Comment by fjl** [comment]
My suggestion for the API is adding a parameter to `JSRE.Stop`:

```go
func (*JSRE) Stop(waitForCallbacks bool)
```

**Comment by zsfelfoldi** [comment]
Fixed the above issues, thanks for the review.

**Comment by fjl** [comment]
Please submit an issue to web3.js. It should be fixed to remove timers when there is nothing to poll.

File: jsre/jsre.go [hunk 11]
```diff
@@ -20,21 +22,174 @@ It provides some helper functions to
...
+
 	return re
 }
 
+func (self *JSRE) run(src interface{}) (value otto.Value, err error) {
+
+	if self.evalRunning {
+		return self.vm.Run(src) // in case of recursive call like loadScript()
+	}
+	self.evalRunning = true
```
Comment (fjl) [inline jsre/jsre.go]: Please change `loadScript` to use the vm directly instead.

**Comment by fjl** [comment]
Note that having the loop means that `self.vm` cannot be used from another goroutine. All calls that use the VM need to go through the `evalQueue` instead. You can address this by changing `evalReq` so it carries a `func()` instead.

**Comment by fjl** [comment]
A few overall comments:

Please at least document `Stop` and its new parameter. Documentation comments generally use the following form:

```go
// Method does something and is very awesome.
func (MyCoolType) Method() foo {

}
```

**Comment by obscuren** [comment]
Accidental label removal

**Comment by zsfelfoldi** [comment]
https://github.com/ethereum/web3.js/issues/184 is also needed for the js filters.

**Comment by fjl** [comment]
What is the status of this PR? Are you still planning to address the issues above?

**Comment by zsfelfoldi** [comment]
Fixed the above issues today.

File: jsre/jsre.go [hunk 12]
```diff
@@ -20,51 +21,233 @@ It provides some helper functions to
...
+func (self *JSRE) Stop(waitForCallbacks bool) {
+	self.stopEventLoop <- waitForCallbacks
+	self.loopWg.Wait()
+}
+
 // Exec(file) loads and runs the contents of a file
 // if a relative path is given, the jsre's assetPath is used
-func (self *JSRE) Exec(file string) error {
-	return self.exec(common.AbsolutePath(self.assetPath, file))
+func (self *JSRE) Exec(file string, useEQ bool) error {
```
Comment (obscuren) [inline jsre/jsre.go]: Exported functions like `Exec` and `Set`/`Get` should not expose the bypassing functionality of the eval loop. There need to be two function:

* `exec(file string, useEQ bool) error`
* `Exec(file string) error`

`Exec` Will always use the eval loop where as `exec` it's optional (and required)

**Comment by obscuren** [comment]
Exported functions need to be documented

**Comment by fjl** [comment]
I believe that the parameter was added so Go functions called from JS can bypass the event loop since
they already run from inside the loop. Sending to the eval channel would hang in those functions.

As an idea for solving this, maybe `Bind` should be modified to pass the `*otto.VM` as the first parameter
to Go functions. The VM could then be used directly inside of the callback.

```go
re := jsre.New(...)
re.Bind("foo", func (vm *otto.VM, call otto.FunctionCall) otto.Value {
    ...  
})
```

**Comment by fjl** [comment]
Looking further, `otto.FunctionCall` already contains the VM as a public field.
I would like to request that the `WithoutEQ` versions are removed and binding functions use
the VM from the call struct instead.

**Comment by fjl** [comment]
This is looking much better now. Thank you. There are minor details that could be cleaned up
but we can do that later.

**Pull request merged**